### ISSUE TITLE
[tf.estimator] Training with tf.estimator + tf.keras and tf.keras only yields inconsistent results

### ISSUE THREAD

**Issue by sdll** [tensorflow/tensorflow#31928]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): `macOS 10.14.6 (18G87)`
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: **n/a**
- TensorFlow installed from (source or binary): **binary**
- TensorFlow version (use command below): `v1.14.0-rc1-22-gaf24dc91b5 1.14.0`
- Python version: `3.7.3`
- Bazel version (if compiling from source): **n/a**
- GCC/Compiler version (if compiling from source): **n/a**
- CUDA/cuDNN version: **n/a**
- GPU model and memory: **none**

**Describe the current behavior**

Following the official guides ([[1]](https://www.tensorflow.org/guide/estimators#creating_estimators_from_keras_models), [[2]](https://www.tensorflow.org/beta/guide/migration_guide#using_a_custom_model_fn)), I was training [PSENet](https://github.com/sdll/psenet/tree/master/psenet) for text detection. Even though train metrics did improve to almost perfect levels and the loss remained stable and low after a while, the inference I got was gibberish.

PSENet works as follows: running the image through the feature pyramid network to obtain segmentation maps, it then applies a custom algorithm to extract bboxes. The images below show the segmentation maps, with yellow regions corresponding to the predicted text, and purple to everything else.

After 186 attempts to make it work on the AI Platform and $400 of GSoC credits, I realized that the problem was deeper than the implementation details, and decided to overfit on a single sample, using the `tf.keras` implementation of FPN from [`segmentation_models`](https://github.com/qubvel/segmentation_models) by @qubvel. I have tweaked his implementation for PSENet and ran into the same problems with `tf.estimator`, so it seems that `tf.estimator` is indeed the culprit.

For this sample image

![sample image](https://raw.githubusercontent.com/sdll/tf.estimator-failing-example/master/docs/input.png)
 
and one of the labels

![sample label](https://raw.githubusercontent.com/sdll/tf.estimator-failing-example/master/docs/label.png)

after 300 epochs, the same loss and optimizer, the predicted labels

- with a pure `tf.keras` implementation:

![](https://raw.githubusercontent.com/sdll/tf.estimator-failing-example/master/docs/training-with-pure-keras.png)

- with the `tf.keras` model converted using `tf.keras.estimator.model_to_estimator`:

![](https://raw.githubusercontent.com/sdll/tf.estimator-failing-example/master/docs/training-with-estimator-and-conversion.png)

- with the `tf.keras` model used in the `tf.estimator` model function:

![](https://raw.githubusercontent.com/sdll/tf.estimator-failing-example/master/docs/training-with-estimator-and-no-conversion.png)

I have tried the `tf.keras`-in-`model_fn` setup on 10 000 images for 30-50 epochs, and the results are much worse than this, which is itself not perfect.

**Describe the expected behavior**

1. TensorFlow documentation should state clearly the preferred way to use `tf.keras` models inside `tf.estimator`, given the knowledge that `tf.estimator` is built on `tf.keras.layers` and thus the expectation that interops is seamless.

2. The discrepancy between training with `tf.keras` and `tf.estimator` + `tf.keras` should be minimal or non-existent

**Code to reproduce the issue**

The minimal failing example with the code and data is [here](https://github.com/sdll/tf.estimator-failing-example).

**Other info / logs**

https://github.com/tensorflow/tensorflow/issues/25670 is a related issue. Similar findings are documented [here](https://stackoverflow.com/questions/54910215/tensorflow-estimator-fails-to-converge-on-model-converted-from-keras-when-using).

I can also confirm that training with `tf.estimator` took longer than with pure `tf.keras`, in alignment with [other reports](https://stackoverflow.com/questions/56930892/tf-estimator-vs-tf-keras-speed-disparity) of this behavior.  

The model itself is sensible, and the following example shows that it does generalize well. For this input (not in the original data):

![IMAGE 2019-08-23 17:13:30](https://user-images.githubusercontent.com/17913919/63598985-549a2f80-c5c9-11e9-9cfe-2db8e72694f9.jpg)

- the output from [another implementation](https://github.com/liuheng92/tensorflow_PSENet) written in tf.slim is as follows:

![IMAGE 2019-08-23 17:14:20](https://user-images.githubusercontent.com/17913919/63599060-72679480-c5c9-11e9-84e1-0b95a9f12edf.jpg)

- the output from the [Pytorch implementation](https://github.com/whai362/PSENet) by the original authors is this:

![IMAGE 2019-08-23 17:16:54](https://user-images.githubusercontent.com/17913919/63599211-cecab400-c5c9-11e9-95e7-e8b48a5cc2e3.jpg)

**Comment by askingbot** [comment]
Seems that there are many batch normalization layers in segmentation model. Using `keras` model in `estimator` should take of the `update_ops`. Otherwise BN's mean and variance will not update.

```python
update_ops = model.get_updates_for(None) + model.get_updates_for(features)
minimize_op = optimizer.minimize(
      total_loss,
      var_list=model.trainable_variables,
      global_step=tf.compat.v1.train.get_or_create_global_step())
train_op = tf.group(minimize_op, update_ops)
```
or in TF1.x style
```python
update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
....
```

https://www.tensorflow.org/beta/guide/migration_guide#using_a_custom_model_fn

**Comment by sdll** [comment]
Nice, could this be added to the [official guide](https://www.tensorflow.org/guide/estimators#creating_estimators_from_keras_models) on estimators? What can explain the fact that `model_to_estimator` performs so poorly? That `train_and_evaluate` should always be used instead of `train`?

**Comment by gowthamkpr** [comment]
@sdll As it is clear that its not a bug, I will go ahead and close the issue and if you have any specific doubts please create a new issue. Thanks!

**Comment by sdll** [comment]
@gowthamkpr, could you please elaborate on what is clear in this instance? I have used tf 1.14 not tf 2.0. Are you sure the sole problem is in not updating the ops? Do the same reasons explain the fact that tf.keras.estimator.model_to_estimator has not worked as expected as well?

**Comment by gowthamkpr** [comment]
@sdll Can you please create a new issue of clearly referring to what the bug is. Thanks!

**Comment by sdll** [comment]
@gowthamkpr, I would love to tell you what the source of the problem is. To help finding this out, I have made a minimum failing example where the model trained in keras works well, while the model trained with keras inside the model function and the keras model converted to the estimator do not. Unfortunately, I do not have enough knowledge of the tf 1.14 internals to say where the training/inference go awry.

**Comment by Jasonnor** [comment]
Any updates on it?  I spent **lots** of time on tf.estimator and tf.keras, but seems they can't be used together perfectly. 😞

**Comment by Saduf2019** [comment]
@sdll 
Is this still an issue

**Comment by sdll** [comment]
Sorry, I do not have enough bandwidth to run the minimum failing example myself to check this out.

**Comment by Saduf2019** [comment]
@sdll 
an you get the example down to the simplest possible reproducible code so that it is easy for us to pinpoint the issue.

**Comment by sdll** [comment]
No, sorry, but if I had time I would go by the route of extracting the tf.keras code for FPN from https://github.com/qubvel/segmentation_models and then running the [example](https://github.com/sdll/tf.estimator-failing-example) again.

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
